src/core/exec-invoke.c

```diff
@@ -4833,6 +4855,30 @@ int exec_invoke(
                 }
         }
 
+        if (context->private_pids) {
+                if (ns_type_supported(NAMESPACE_PID)) {
+                        PidRef pidref;
+
+                        r = pidref_safe_fork(NULL, FORK_NEW_PIDNS, &pidref);
+                        if (ERRNO_IS_NEG_PRIVILEGE(r))
+                                log_exec_warning_errno(context, params, r,
+                                                       "PrivatePIDs=yes is configured, but pid namespace setup failed, ignoring: %m");
+                        else if (r < 0) {
+                                *exit_status = EXIT_NAMESPACE;
+                                return log_exec_error_errno(context, params, r, "Failed to fork child process into new pid namespace: %m");
+                        }
+                        if (r > 0) {
+                                /* In the parent process, we send the child pidref to the manager and exit. */
+                                r = send_pidref(context, params, &pidref, exit_status);
+                                if (r < 0)
+                                        return r;
+
+                                return 0;
+                        }
```

Comment:
so do i get this right? your fork here and continue the main code in the child rather than the parent?

that's not OK really. because the fork uses raw_clone(), which confuses the hell out of glibc and its mutex use for protecting things such as stdio. if we invoke raw_clone() then we must be extremely conservative about what we invoke in the child. And I am pretty sure just letting the rest of the initialization run here might be highly problematic. for example, are you sure nothing here calls stdio calls after this point?

or am i misunderstanding things?

Reply:
Doesn't nspawn do exactly the same?

What we can consider is splitting this up, unshare here to make sure it happens before or after unsharing userns based on whether we're running privileged or not, but only fork at the very end just before we exec.

Reply:
We now call `unshare(CLONE_NEWPID)` and then do a regular fork. That gets around the issue with `raw_clone()`?

Reply:
yeah, i think the new approach is workable, but please add comments, highlighting the weirdness, see my other review points.